Add ALTS_TARGET_NAME_OVERRIDE_ARG channel arg

apolcyn opened the pull request:
b/156545190 for context

jiangtaoli2016 commented on include/grpc/impl/codegen/grpc_types.h:
```diff
@@ -272,6 +272,14 @@ typedef struct {
     grpc_ssl_session_cache*). (use grpc_ssl_session_cache_arg_vtable() to fetch
     an appropriate pointer arg vtable) */
 #define GRPC_SSL_SESSION_CACHE_ARG "grpc.ssl_session_cache"
+/** The caller of the secure_channel_create functions may override the target
```
I would add a warning here

This should only be used for testing or when a client uses IP address to connect to a service.

apolcyn replied:
done

markdroth commented on src/core/lib/surface/channel.cc:
```diff
@@ -143,10 +144,15 @@ static grpc_core::UniquePtr<char> get_default_authority(
                            GRPC_SSL_TARGET_NAME_OVERRIDE_ARG)) {
       ssl_override = grpc_channel_arg_get_string(&input_args->args[i]);
+    } else if (0 == strcmp(input_args->args[i].key,
+                           GRPC_ALTS_TARGET_NAME_OVERRIDE_ARG)) {
```
I'm really not keen on adding this special case to the channel code itself.  Can we instead set this in the ALTS credentials?

I realize that the corresponding case for SSL creds is here, but I think that's a mistake.  I'd like to see that one moved into SSL creds as well.

apolcyn replied:
Done, I moved both the SSL and ALTS override codes into the channel_credentials [update_arguments](https://github.com/grpc/grpc/blob/83dd71dfe58253b4ceb3cf394a939f70adc7ea95/src/core/lib/security/credentials/credentials.h#L128) hook.

This then orphaned this function in `src/core/lib/surface/channel.cc`.

markdroth commented on src/core/lib/surface/channel.cc:
```diff
@@ -143,10 +144,15 @@ static grpc_core::UniquePtr<char> get_default_authority(
   }
   if (!has_default_authority && ssl_override != nullptr) {
     default_authority.reset(gpr_strdup(ssl_override));
+  } else if (!has_default_authority && alts_override != nullptr) {
```
If the checks above stay here, this should just be combined with the check on line 152.

```
if (!has_default_authority && ssl_override != nullptr && alts_override != nullptr) {
```

apolcyn replied:
I've removed this code from `src/core/lib/surface/channel.cc`

markdroth commented on test/core/tsi/alts/handshaker/alts_concurrent_connectivity_test.cc:
```diff
@@ -105,11 +105,15 @@ class FakeHandshakeServer {
+      service_ = grpc::gcp::CreateFakeHandshakerService(
+          kFakeHandshakeServerMaxConcurrentStreams /* expected max concurrent
+                                                      rpcs */
+          ,
```
The comma should move to before the `/*`.

apolcyn replied:
done, removed comment since the `kFakeHandshakeServerMaxConcurrentStreams` constant seemed descriptive.

markdroth commented on test/core/tsi/alts/handshaker/alts_target_name_override_test.cc:
```diff
@@ -0,0 +1,284 @@
+  cq_verify(cqv);
+  GPR_ASSERT(status == GRPC_STATUS_OK);
+  // Extract the authority header
+  grpc_core::UniquePtr<char> authority_header(
```
Let's not add new uses of `UniquePtr<char>`, since we're trying to get rid of it.

```
std::string authority_header(grpc_core::StringViewFromSlice(call_details.host));
```

apolcyn replied:
done

markdroth commented on test/core/tsi/alts/handshaker/alts_target_name_override_test.cc:
```diff
@@ -0,0 +1,284 @@
+  op->flags = 0;
+  op->reserved = nullptr;
+  op++;
+  op->op = GRPC_OP_SEND_MESSAGE;
```
None of the send_message or recv_message ops are actually needed in this test.

apolcyn replied:
done, removed send and receive message ops

apolcyn commented:
@markdroth I believe I've addressed comments, PTAL

Regarding your concerns about the high-level approach that calls for this PR in the first place, it makes sense - I'm looking at XDS docs and am also planning to create a load balancing "requirements" doc for the customer to get a better handle on things. However, I think I'd still like to pursue my original idea in the background, if nothing else for visibility as a tactical alternative (the only pre-requisite to submitting a working prototype of that idea is this PR).
